econsa/shapley.py

```diff
@@ -1,9 +1,214 @@
+dependent inputs.
 
 """
 
+import itertools
+import numpy as np
+import pandas as pd
+import chaospy as cp
+from econsa.sampling import cond_mvn
+
+
+def get_shapley(
+    method, model, Xall, Xcond, n_perms, n_inputs, n_output, n_outer, n_inner,
+):
+    """
+
+    This function calculates Shapley effects and their standard errors for
+    models with both dependent and independent inputs. We allow for two ways
+    to calculate Shapley effects: by examining all permutations of the given
+    inputs or alternatively, by randomly sampling permutations of inputs.
+
+    The function is a translation of the exact and random permutation funtions
+    in the ``sensitivity`` package in R, and takes the method (exact or random)
+    as an argument and therefore estimates shapley effects in both ways.
+
+
+    Parameters
+    ----------
+    method : string
+           Specifies which method you want to use to estimate shapley effects,
+           the ``exact`` or ``random`` permutations method. When the number of
+           inputs is small, it is better to use the ``exact`` method, and
+           ``random`` otherwise.
+
+    model : string
+          The model/function you will calculate the shapley effects on.
+
+    Xall : string (n)
+         A function that takes `n´ as an argument and generates a n-sample of
+         a d-dimensional input vector.
+
+    Xset : string (n, Sj, Sjc, xjc)
+         A function that takes `n, Sj, Sjc, xjc´ as arguments and generates
+         a n- sample an input vector corresponding to the indices in `Sj´
+         conditional on the input values `xjc´ with the index set `Sjc´.
+
+    n_perms : scalar
+            This is an input for the number of permutations you want the model
+            to make. For the ``exact`` method, this argument is none as the
+            number of permutations is determined by how many inputs you have,
+            and for the ``random`` method, this is determined exogeniously.
+
+    n_inputs : scalar
+             The number of input vectors for which shapley estimates are being
+             estimated.
+
+    n_output : scalar
+             Monte Carlo (MC) sample size to estimate the output variance of
+             the model output `Y´.
+
+    n_outer : scalar
+            The outer Monte Carlo sample size to estimate the cost function for
+            `c(J) = E[Var[Y|X]]´.
+
+    n_inner : scalar
+            The inner Monte Carlo sample size to estimate the cost function for
+            `c(J) = Var[Y|X]´.
+
+    Returns
+    -------
+    effects : DataFrame
+            n dimensional DataFrame with the estimated shapley effects, the
+            standard errors and the confidence intervals for the input vectors.
+
+    Sources
+    -------
+    Shapley exact permutations: https://rdrr.io/cran/sensitivity/src/R/shapleyPermEx.R
+    Shapley random permutations: https://rdrr.io/cran/sensitivity/src/R/shapleyPermRand.R
```

Comment:
Please use proper ReST links or references, check other functions, temfpy, or [reStructuredText Markup Specification](https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html) if you’re not familiar with that.

Reply:
done